File: frontend/src/modules/finance/api/purchasePlans.ts
/**
 * Typed REST client for /finance/purchase-plans.
 *
 * A PurchasePlan represents a single purchase materialized across N
 * installment transactions. The aggregate itself is not a ledger: the
 * backend spawns one Transaction per installment under the plan's id.
 * Future installments arrive as `scheduled`, paid ones as `paid` —
 * which means every total flows through the existing transactions
 * pipeline without a plan-specific aggregation. See
 * `backend/docs/totals-contract.md` §3.
 *
 * Wire shape mirrors `backend/api/openapi/finance.yaml` and the Go
 * structs in `backend/internal/finance/adapters/httpapi/purchaseplans.go`
 * + `backend/internal/finance/domain/purchaseplan.go`:
 *   • snake_case fields, integer cents, RFC3339 timestamps
 *   • create body keys: name / total_amount_cents / installments /
 *     first_occurred_at / category_id / person_id / account_id /
 *     payment_method / source / notes
 *   • response keys: id / workspace_id / person_id / name /
 *     total_amount_cents / installments / remaining_installments /
 *     created_at / deleted_at
 */

import { apiFetch, type PageEnvelope } from "@/lib/api/client";
import type {
  PaymentMethodWire,
  TransactionSourceWire,
} from "./transactions";

export interface ApiPurchasePlan {
  id: string;
  workspace_id: string;
  person_id: string | null;
  name: string;
  total_amount_cents: number;
  installments: number;
  remaining_installments: number;
  created_at: string;            // RFC3339
  deleted_at?: string | null;
}

export interface CreatePurchasePlanRequest {
  category_id: string;
  person_id?: string | null;
  // FK deferred on the backend side (migration 0004 comment); the FE
  // passes the chosen credit-card UUID here so each installment row
  // carries it back via `transaction.account_id`.
  account_id?: string | null;
  name: string;
  total_amount_cents: number;
  installments: number;
  first_occurred_at: string;     // RFC3339
  payment_method?: PaymentMethodWire;
  source?: TransactionSourceWire;
  notes?: string;
}

export interface ListPurchasePlansParams {
  limit?: number;
  offset?: number;
}

function qs(params: object): string {
  const sp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined) sp.set(k, String(v));
  }
  const s = sp.toString();
  return s ? `?${s}` : "";
}

export async function listPurchasePlans(
  p: ListPurchasePlansParams = {},
): Promise<ApiPurchasePlan[]> {
  const env = await apiFetch<PageEnvelope<ApiPurchasePlan>>(
    `/finance/purchase-plans${qs({ limit: 100, ...p })}`,
  );
  return env.items;
}

export function getPurchasePlan(id: string): Promise<ApiPurchasePlan> {
  return apiFetch<ApiPurchasePlan>(`/finance/purchase-plans/${id}`);
}

export function createPurchasePlan(
  body: CreatePurchasePlanRequest,
): Promise<ApiPurchasePlan> {
  return apiFetch<ApiPurchasePlan>(`/finance/purchase-plans`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

// Backend cancel is body-less: it soft-deletes every still-scheduled
// installment of the plan. Paid installments stay as history.
export function cancelPurchasePlan(id: string): Promise<ApiPurchasePlan> {
  return apiFetch<ApiPurchasePlan>(`/finance/purchase-plans/${id}/cancel`, {
    method: "POST",
  });
}

File: backend/internal/finance/adapters/httpapi/purchaseplans.go
package httpapi

import (
	"net/http"
	"time"

	"github.com/google/uuid"

	"github.com/corsi/backend/internal/finance/app"
	"github.com/corsi/backend/internal/finance/domain"
	"github.com/corsi/backend/internal/platform/apierror"
	"github.com/corsi/backend/internal/platform/render"
)

// createPurchasePlanReq is the wire shape for POST /purchase-plans. The
// handler hands it straight to Service.CreatePurchasePlan, which writes
// the parent plan row and the N installment transactions inside one tx
// — see the service comment for rollback semantics.
type createPurchasePlanReq struct {
	CategoryID       uuid.UUID                 `json:"category_id"`
	PersonID         *uuid.UUID                `json:"person_id,omitempty"`
	AccountID        *uuid.UUID                `json:"account_id,omitempty"`
	Name             string                    `json:"name"`
	TotalAmountCents int64                     `json:"total_amount_cents"`
	Installments     int                       `json:"installments"`
	FirstOccurredAt  time.Time                 `json:"first_occurred_at"`
	PaymentMethod    *domain.PaymentMethod     `json:"payment_method,omitempty"`
	Source           *domain.TransactionSource `json:"source,omitempty"`
	Notes            string                    `json:"notes,omitempty"`
}

func (h *Handler) createPurchasePlan(w http.ResponseWriter, r *http.Request) {
	wsID, ok := h.workspaceID(w, r)
	if !ok {
		return
	}
	var req createPurchasePlanReq
	if err := render.DecodeJSON(r, &req); err != nil {
		apierror.Write(w, h.log, apierror.New(http.StatusBadRequest, "malformed_body", err.Error()))
		return
	}
	p, err := h.svc.CreatePurchasePlan(r.Context(), app.CreatePurchasePlanInput{
		WorkspaceID:      wsID,
		PersonID:         req.PersonID,
		CategoryID:       req.CategoryID,
		AccountID:        req.AccountID,
		Name:             req.Name,
		TotalAmountCents: req.TotalAmountCents,
		Installments:     req.Installments,
		FirstOccurredAt:  req.FirstOccurredAt,
		PaymentMethod:    req.PaymentMethod,
		Source:           req.Source,
		Notes:            req.Notes,
	})
	if err != nil {
		h.writeDomainErr(w, err)
		return
	}
	render.JSON(w, http.StatusCreated, p)
}

func (h *Handler) getPurchasePlan(w http.ResponseWriter, r *http.Request) {
	wsID, ok := h.workspaceID(w, r)
	if !ok {
		return
	}
	id, err := parseUUIDParam(r, "id")
	if err != nil {
		apierror.Write(w, h.log, apierror.New(http.StatusBadRequest, "invalid_id", "id must be a uuid"))
		return
	}
	p, err := h.svc.GetPurchasePlan(r.Context(), wsID, id)
	if err != nil {
		h.writeDomainErr(w, err)
		return
	}
	render.JSON(w, http.StatusOK, p)
}

func (h *Handler) listPurchasePlans(w http.ResponseWriter, r *http.Request) {
	wsID, ok := h.workspaceID(w, r)
	if !ok {
		return
	}
	limit, offset := parseLimitOffset(r)
	items, err := h.svc.ListPurchasePlans(r.Context(), app.ListPurchasePlansInput{
		WorkspaceID: wsID, Limit: limit, Offset: offset,
	})
	if err != nil {
		h.writeDomainErr(w, err)
		return
	}
	render.JSON(w, http.StatusOK, listEnvelope{Items: items, Limit: limit, Offset: offset})
}

func (h *Handler) cancelPurchasePlan(w http.ResponseWriter, r *http.Request) {
	wsID, ok := h.workspaceID(w, r)
	if !ok {
		return
	}
	id, err := parseUUIDParam(r, "id")
	if err != nil {
		apierror.Write(w, h.log, apierror.New(http.StatusBadRequest, "invalid_id", "id must be a uuid"))
		return
	}
	p, err := h.svc.CancelPurchasePlan(r.Context(), wsID, id)
	if err != nil {
		h.writeDomainErr(w, err)
		return
	}
	render.JSON(w, http.StatusOK, p)
}
